File: apps/backend/app/schemas/invoice_schema.py
from pydantic import BaseModel
from typing import Optional, List


class VendorSchema(BaseModel):
    name: Optional[str] = None
    gstin: Optional[str] = None
    pan: Optional[str] = None
    address: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None


class CustomerSchema(BaseModel):
    name: Optional[str] = None
    gstin: Optional[str] = None
    address: Optional[str] = None


class LineItemSchema(BaseModel):
    description: Optional[str] = None
    hsn_sac: Optional[str] = None
    quantity: Optional[float] = None
    unit: Optional[str] = None
    unit_price: Optional[float] = None
    taxable_value: Optional[float] = None
    cgst_rate: Optional[float] = None
    cgst_amount: Optional[float] = None
    sgst_rate: Optional[float] = None
    sgst_amount: Optional[float] = None
    igst_rate: Optional[float] = None
    igst_amount: Optional[float] = None
    total: Optional[float] = None


class TaxSummarySchema(BaseModel):
    subtotal: Optional[float] = None
    cgst_total: Optional[float] = None
    sgst_total: Optional[float] = None
    igst_total: Optional[float] = None
    total_tax: Optional[float] = None
    round_off: Optional[float] = None
    grand_total: Optional[float] = None


class BankDetailsSchema(BaseModel):
    account_name: Optional[str] = None
    account_number: Optional[str] = None
    bank_name: Optional[str] = None
    ifsc: Optional[str] = None
    branch: Optional[str] = None


class InvoiceDataSchema(BaseModel):
    invoice_number: Optional[str] = None
    invoice_date: Optional[str] = None
    due_date: Optional[str] = None
    currency: Optional[str] = None
    vendor: VendorSchema = VendorSchema()
    customer: CustomerSchema = CustomerSchema()
    purchase_order_number: Optional[str] = None
    line_items: List[LineItemSchema] = []
    tax_summary: TaxSummarySchema = TaxSummarySchema()
    bank_details: BankDetailsSchema = BankDetailsSchema()


class ConfidenceSchema(BaseModel):
    overall: float = 0.0
    invoice_number: float = 0.0
    invoice_date: float = 0.0
    gstin: float = 0.0
    line_items: float = 0.0
    totals: float = 0.0


class RuleCheckSchema(BaseModel):
    rule: str
    passed: bool
    message: str
    field: Optional[str] = None


class LLMCheckSchema(BaseModel):
    check: str
    result: str
    confidence: float
    message: str
    field: Optional[str] = None


class ValidationResultSchema(BaseModel):
    status: str
    rule_checks: List[RuleCheckSchema] = []
    llm_checks: List[LLMCheckSchema] = []
    warnings: List[str] = []
    errors: List[str] = []


class MetadataSchema(BaseModel):
    ocr_engine: Optional[str] = None
    complexity_score: Optional[float] = None
    processing_mode: Optional[str] = None
    pages: int = 0


class InvoiceOutputSchema(BaseModel):
    document_id: str
    invoice: InvoiceDataSchema = InvoiceDataSchema()
    confidence: ConfidenceSchema = ConfidenceSchema()
    validation: ValidationResultSchema = ValidationResultSchema(status="PENDING")
    metadata: MetadataSchema = MetadataSchema()

File: mastra-service/src/mastra/schemas/invoiceSchema.ts
import { z } from "zod";

export const VendorSchema = z.object({
  name: z.string().nullable().optional(),
  gstin: z.string().nullable().optional(),
  pan: z.string().nullable().optional(),
  address: z.string().nullable().optional(),
  phone: z.string().nullable().optional(),
  email: z.string().nullable().optional(),
});

export const CustomerSchema = z.object({
  name: z.string().nullable().optional(),
  gstin: z.string().nullable().optional(),
  address: z.string().nullable().optional(),
});

export const LineItemSchema = z.object({
  description: z.string().nullable().optional(),
  hsn_sac: z.string().nullable().optional(),
  quantity: z.number().nullable().optional(),
  unit: z.string().nullable().optional(),
  unit_price: z.number().nullable().optional(),
  taxable_value: z.number().nullable().optional(),
  cgst_rate: z.number().nullable().optional(),
  cgst_amount: z.number().nullable().optional(),
  sgst_rate: z.number().nullable().optional(),
  sgst_amount: z.number().nullable().optional(),
  igst_rate: z.number().nullable().optional(),
  igst_amount: z.number().nullable().optional(),
  total: z.number().nullable().optional(),
});

export const TaxSummarySchema = z.object({
  subtotal: z.number().nullable().optional(),
  cgst_total: z.number().nullable().optional(),
  sgst_total: z.number().nullable().optional(),
  igst_total: z.number().nullable().optional(),
  total_tax: z.number().nullable().optional(),
  round_off: z.number().nullable().optional(),
  grand_total: z.number().nullable().optional(),
});

export const BankDetailsSchema = z.object({
  account_name: z.string().nullable().optional(),
  account_number: z.string().nullable().optional(),
  bank_name: z.string().nullable().optional(),
  ifsc: z.string().nullable().optional(),
  branch: z.string().nullable().optional(),
});

export const InvoiceDataSchema = z.object({
  invoice_number: z.string().nullable().optional(),
  invoice_date: z.string().nullable().optional(),
  due_date: z.string().nullable().optional(),
  currency: z.string().nullable().optional(),
  vendor: VendorSchema.optional().default({}),
  customer: CustomerSchema.optional().default({}),
  purchase_order_number: z.string().nullable().optional(),
  line_items: z.array(LineItemSchema).optional().default([]),
  tax_summary: TaxSummarySchema.optional().default({}),
  bank_details: BankDetailsSchema.optional().default({}),
});

export const ConfidenceSchema = z.object({
  overall: z.number().default(0),
  invoice_number: z.number().default(0),
  invoice_date: z.number().default(0),
  gstin: z.number().default(0),
  line_items: z.number().default(0),
  totals: z.number().default(0),
});

export const RuleCheckSchema = z.object({
  rule: z.string(),
  passed: z.boolean(),
  message: z.string(),
  field: z.string().nullable().optional(),
});

export const LLMCheckSchema = z.object({
  check: z.string(),
  result: z.enum(["PASS", "FAIL", "UNCERTAIN"]),
  confidence: z.number(),
  message: z.string(),
  field: z.string().nullable().optional(),
});

export const ValidationSchema = z.object({
  status: z.enum(["VALID", "NEEDS_REVIEW", "INVALID", "PENDING"]),
  rule_checks: z.array(RuleCheckSchema).default([]),
  llm_checks: z.array(LLMCheckSchema).default([]),
  warnings: z.array(z.string()).default([]),
  errors: z.array(z.string()).default([]),
});

export const MetadataSchema = z.object({
  ocr_engine: z.string().nullable().optional(),
  complexity_score: z.number().nullable().optional(),
  processing_mode: z.string().nullable().optional(),
  pages: z.number().default(0),
});

export const InvoiceOutputSchema = z.object({
  document_id: z.string(),
  invoice: InvoiceDataSchema.default({}),
  confidence: ConfidenceSchema.default({ overall: 0, invoice_number: 0, invoice_date: 0, gstin: 0, line_items: 0, totals: 0 }),
  validation: ValidationSchema.default({ status: "PENDING", rule_checks: [], llm_checks: [], warnings: [], errors: [] }),
  metadata: MetadataSchema.default({ pages: 0 }),
});

export type InvoiceOutput = z.infer<typeof InvoiceOutputSchema>;
export type InvoiceData = z.infer<typeof InvoiceDataSchema>;
export type LineItem = z.infer<typeof LineItemSchema>;
